src/core/utils/i18n.js

```diff
@@ -39,3 +40,83 @@ export const getLocalizedTextWithLinkParts = ({
     afterLinkText: parts[2],
   };
 };
+
+const getReplacementKey = (start: string, end: string): string => {
+  return `${start},${end}`;
+};
+
+type ReplaceStringsWithJSXParams = {|
+  text: string,
+  replacements: Array<[string, string, (text: string) => React.Node]>,
+|};
+
+export const replaceStringsWithJSX = ({
+  text,
+  replacements,
+}: ReplaceStringsWithJSXParams): Array<React.Node | string> => {
+  if (replacements.length === 0) {
+    return [text];
+  }
+
+  const expression = new RegExp(
+    [
+      // Before any replacement (pair of keys).
+      '^(.*?)',
+      // Add a generic regexp to match each replacement entry, possibly
+      // separated by some more text.
+      replacements.map(() => '%\\((\\w+)\\)s(.+?)%\\((\\w+)\\)s').join('(.*?)'),
+      // After all replacements (pairs of keys).
+      '(.*?)$',
+    ].join(''),
+  );
+
+  const matches = text.match(expression);
+  const internalReplacements = replacements.reduce((map, replacement) => {
+    return {
+      ...map,
+      [getReplacementKey(replacement[0], replacement[1])]: replacement[2],
+    };
+  }, {});
+
+  if (!matches) {
+    // This usually happens when the number of pairs of keys in `text` is not
+    // the same as the number of entries in the `replacements` argument.
+    throw new Error(
+      '`text` does not appear to be compatible with the provided replacements',
+    );
+  }
+
+  // The first entry is always the whole string (a.k.a. `text`), so let's
+  // remove it before processing this array.
+  matches.shift();
+
+  const output = [];
+  while (matches.length) {
+    const keyOrText = matches.shift();
+
+    // Look-ahead to see if we found a known pair of keys.
+    if (internalReplacements[getReplacementKey(keyOrText, matches[1])]) {
+      const innerText = matches.shift();
+      const secondKey = matches.shift();
+
+      const key = getReplacementKey(keyOrText, secondKey);
+      const replaceFn = internalReplacements[key];
+
+      output.push(replaceFn(innerText));
+      // eslint-disable-next-line no-param-reassign
+      delete internalReplacements[key];
+    } else {
```

Comment:
If no replacement is given for a format string, it looks pretty bad. Can you think of a way to detect that and throw an error? I think it would be safe to throw an error because we have dennis to catch L10n mistakes for this. In other words, it would only be a programming error, not a localizer error.

Here's a test I used to see what would happen:

```diff
diff --git a/tests/unit/core/utils/test_i18n.js b/tests/unit/core/utils/test_i18n.js
index 83978b035..9a0dfb4e3 100644
--- a/tests/unit/core/utils/test_i18n.js
+++ b/tests/unit/core/utils/test_i18n.js
@@ -117,6 +117,26 @@ describe(__filename, () => {
       ]);
     });
 
+    it('throws an error for unmatched format strings', () => {
+      expect(
+        replaceStringsWithJSX({
+          text:
+            'Click on %(redLinkStart)sred%(redLinkEnd)s or %(blueLinkStart)sblue%(blueLinkEnd)s, your choice',
+          replacements: [
+            [
+              'redLinkStart',
+              'redLinkEnd',
+              (text) => (
+                <Link key="red" to="/red">
+                  {text}
+                </Link>
+              ),
+            ],
+          ],
+        }),
+      ).toEqual(['...']);
+    });
+
     it('lets you replace format strings with JSX in any order', () => {
       expect(
         replaceStringsWithJSX({
```

```
    Difference:

    - Expected
    + Received

      Array [
    -   "...",
    +   "Click on ",
    +   <Connect(LinkBase)
    +     to="/red"
    +   >
    +     red
    +   </Connect(LinkBase)>,
    +   " or %(blueLinkStart)sblue%(blueLinkEnd)s, your choice",
```

Reply:
Isn't it caught by the `Not all replacements have been used` error?

Reply:
I added some code to make sure we have the same number of replacements and placeholders in `text`.